crates/utils/src/settings/mod.rs

```diff
@@ -0,0 +1,219 @@
+use crate::{
+  location_info,
+  settings::{
+    captcha_config::CaptchaConfig,
+    database_config::DatabaseConfig,
+    email_config::EmailConfig,
+    federation_config::FederationConfig,
+    rate_limit_config::RateLimitConfig,
+    setup_config::SetupConfig,
+  },
+  LemmyError,
+};
+use anyhow::{anyhow, Context};
+use deser_hjson::from_str;
+use merge::Merge;
+use serde::Deserialize;
+use std::{
+  env,
+  fs,
+  io::Error,
+  net::{IpAddr, Ipv4Addr},
+  sync::RwLock,
+};
+
+pub(crate) mod captcha_config;
+pub(crate) mod database_config;
+pub(crate) mod email_config;
+pub(crate) mod federation_config;
+pub(crate) mod rate_limit_config;
+pub(crate) mod setup_config;
+
+static CONFIG_FILE: &str = "config/config.hjson";
+
+#[derive(Debug, Deserialize, Clone, Merge)]
+pub struct Settings {
+  database: Option<DatabaseConfig>,
+  rate_limit: Option<RateLimitConfig>,
+  federation: Option<FederationConfig>,
+  hostname: Option<String>,
+  bind: Option<IpAddr>,
+  port: Option<u16>,
+  tls_enabled: Option<bool>,
+  jwt_secret: Option<String>,
+  pictrs_url: Option<String>,
+  iframely_url: Option<String>,
+  captcha: Option<CaptchaConfig>,
+  email: Option<EmailConfig>,
+  setup: Option<SetupConfig>,
+}
+
+impl Default for Settings {
+  fn default() -> Self {
+    Self {
+      database: Some(DatabaseConfig::default()),
+      rate_limit: Some(RateLimitConfig::default()),
+      federation: Some(FederationConfig::default()),
+      captcha: Some(CaptchaConfig::default()),
+      email: None,
+      setup: None,
+      hostname: None,
+      bind: Some(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0))),
+      port: Some(8536),
+      tls_enabled: Some(true),
+      jwt_secret: Some("changeme".into()),
+      pictrs_url: Some("http://pictrs:8080".into()),
+      iframely_url: Some("http://iframely".into()),
+    }
+  }
+}
+
+lazy_static! {
+  static ref SETTINGS: RwLock<Settings> = RwLock::new(match Settings::init() {
+    Ok(c) => c,
+    Err(e) => panic!("{}", e),
+  });
+}
+
+impl Settings {
+  /// Reads config from the files and environment.
+  /// First, defaults are loaded from CONFIG_FILE_DEFAULTS, then these values can be overwritten
+  /// from CONFIG_FILE (optional). Finally, values from the environment (with prefix LEMMY) are
+  /// added to the config.
+  ///
+  /// Note: The env var `LEMMY_DATABASE_URL` is parsed in
+  /// `lemmy_db_queries/src/lib.rs::get_database_url_from_env()`
+  fn init() -> Result<Self, LemmyError> {
+    // Read the config file
+    let mut custom_config = from_str::<Settings>(&Self::read_config_file()?)?;
+
+    // Merge with env vars
+    custom_config.merge(envy::prefixed("LEMMY_").from_env::<Settings>()?);
```

Comment:
Okay @Nutomic , got env vars working. I took this from your PR, and it worked (see below)

Reply:
So it works for all config values, including nested ones? Can we revert the changes to configuration in `api_tests/` and `docker/` then?

Reply:
We *could*, but since I already wrote the configs as proper `hjson` files, I don't think there's a reason to.